server/sonar-server/src/main/java/org/sonar/server/computation/period/PeriodFinder.java

```diff
@@ -0,0 +1,125 @@
+    if (snapshot == null) {
+      return null;
+    }
+    return new Period(CoreProperties.TIMEMACHINE_MODE_DATE, date.getTime(), snapshot).setModeParameter(DATE_FORMAT.format(date));
+  }
+
+  @CheckForNull
+  Period findByDays(DbSession session, Long projectId, long analysisDate, int days) {
+    List<SnapshotDto> snapshots = dbClient.snapshotDao().selectSnapshotsByQuery(session, createCommonQuery(projectId).setCreatedBefore(analysisDate).setSort(BY_DATE, ASC));
+    long targetDate = DateUtils.addDays(new Date(analysisDate), -days).getTime();
+    SnapshotDto snapshot = findNearestSnapshotToTargetDate(snapshots, targetDate);
+    if (snapshot == null) {
+      return null;
+    }
+    return new Period(CoreProperties.TIMEMACHINE_MODE_DAYS, targetDate, snapshot).setModeParameter(String.valueOf(days));
+  }
+
+  @CheckForNull
+  Period findByPreviousAnalysis(DbSession session, Long projectId, long analysisDate) {
+    SnapshotDto snapshot = findFirstSnapshot(session, createCommonQuery(projectId).setCreatedBefore(analysisDate).setIsLast(true).setSort(BY_DATE, DESC));
+    if (snapshot == null) {
+      return null;
+    }
+    return new Period(CoreProperties.TIMEMACHINE_MODE_PREVIOUS_ANALYSIS, snapshot.getCreatedAt(), snapshot).setModeParameter(DATE_FORMAT.format(snapshot.getCreatedAt()));
+  }
+
+  @CheckForNull
+  Period findByPreviousVersion(DbSession session, Long projectId, String version) {
+    List<SnapshotDto> snapshotDtos = dbClient.snapshotDao().selectPreviousVersionSnapshots(session, projectId, version);
+    if (snapshotDtos.isEmpty()) {
+      return null;
+    }
+    SnapshotDto snapshotDto = snapshotDtos.get(0);
+    return new Period(CoreProperties.TIMEMACHINE_MODE_PREVIOUS_VERSION, snapshotDto.getCreatedAt(), snapshotDto).setModeParameter(snapshotDto.getVersion());
+  }
+
+  @CheckForNull
+  Period findByVersion(DbSession session, Long projectId, String version) {
+    SnapshotDto snapshot = findFirstSnapshot(session, createCommonQuery(projectId).setVersion(version).setSort(BY_DATE, DESC));
+    if (snapshot == null) {
+      return null;
+    }
+    return new Period(CoreProperties.TIMEMACHINE_MODE_VERSION, snapshot.getCreatedAt(), snapshot).setModeParameter(version);
+  }
+
+  @VisibleForTesting
+  @CheckForNull
+  static SnapshotDto findNearestSnapshotToTargetDate(List<SnapshotDto> snapshots, Long targetDate) {
+    long bestDistance = Long.MAX_VALUE;
+    SnapshotDto nearest = null;
+    for (SnapshotDto snapshot : snapshots) {
+      long distance = Math.abs(snapshot.getCreatedAt() - targetDate);
+      if (distance <= bestDistance) {
+        bestDistance = distance;
+        nearest = snapshot;
+      }
+    }
+    return nearest;
+  }
+
+  @CheckForNull
+  private SnapshotDto findFirstSnapshot(DbSession session, SnapshotQuery query) {
+    List<SnapshotDto> snapshots = dbClient.snapshotDao().selectSnapshotsByQuery(session, query);
+    if (snapshots.size() >= 1) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use isEmpty() to check whether the collection is empty or not. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1155)